services/horizon: Parallelize `db reingest range`

2opremio opened the pull request:
<!-- If you're making a doc PR or something tiny where the below is irrelevant, delete this
template and use a short description, but in your description aim to include both what the
change is, and why it is being made, with enough context for anyone to understand. -->

<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

This change breaks down the ledger range to reingest in subranges which are submitted to a pre-defined number of workers, processing the subranges in parallel.

For now, the workers are simply Go routines using their own `System` (with their own DB connections etc ...).

In the future workers could be fully fledged Horizon instances running in multiple machines (e.g. orchestrated through Kubernetes Jobs or AWS Batch Jobs).

New flags:
--parallel-workers: [optional] if this flag is set to > 1, horizon will parallelize reingestion using the supplied number of workers
--parallel-job-size: [optional] parallel workers will run jobs processing ledger batches of the supplied size

### Why

We want reingestion to be faster. Addresses #2552 

### Known limitations

As I mentioned, this change only applies to a single machine. Running it in multiple machines will require further work.

This PR is targetting `release-1.5.0`  (since that's the base branch I used). We should retarget it to master (once `release-1.5.0` is merged) before merging.

bartekn commented on services/horizon/internal/expingest/main.go:
```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+func (ps *ParallelSystems) ReingestRange(fromLedger, toLedger uint32, batchSizeSuggestion uint32) error {
+	batchSize := calculateParallelLedgerBatchSize(toLedger-fromLedger, batchSizeSuggestion, ps.workerCount)
+	pendingJobsCount := 0
+	var result error
```
I think `err` better describes this variable.
```suggestion
	var err error
```

bartekn commented on services/horizon/internal/expingest/main.go:
```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+			processSubRangeResult(subRangeResult)
+		case ps.reingestJobQueue <- subRange:
+			pendingJobsCount++
+			subRangeFrom = subRangeTo
```
Don't we ingest last ledger in each subrange twice now?
```suggestion
			subRangeFrom = subRangeTo+1
```

2opremio replied:
Oh, that is true, good catch ... although ... shouldn't we start and end in checkpoint limits? (I got _inspiration_ from verify-range which I think uses the same approach I use here).

bartekn replied:
This isn't necessary when reingesting history. We need to start and end with a checkpoint ledger in verify range because we need to build and then verify the state. Since history ingestion does not update state it's not needed.

bartekn commented on services/horizon/internal/expingest/main.go:
```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+			return errors.New("aborted")
+		case subRangeResult := <-ps.reingestJobResult:
+			processSubRangeResult(subRangeResult)
+		case ps.reingestJobQueue <- subRange:
```
It was hard to me to understand the code because subrange calculations are happening even when not needed. Maybe we can move this to a one-liner like:
```suggestion
		case ps.reingestJobQueue <- ledgerRange{from: subRangeFrom, to: min(subRangeFrom + batchSize, toLedger)}:
```

2opremio replied:
Well, that doesn't really work because `min()` doesn't exist and you need to use `min(subRangeFrom+batchSize, toLedger)` in the body of the `case` (`subRangeTo`). So, I don't think that a real solution makes it easier to read.

bartekn replied:
Yes, the assumption was to add `min` function but if you are planning to do https://github.com/stellar/go/pull/2724#discussion_r444326321 we can ignore this comment.

2opremio replied:
After the refactoring you suggested below, it's closer to what you suggest.

bartekn commented on services/horizon/internal/expingest/main.go:
```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+			processSubRangeResult(subRangeResult)
+		}
+
+	}
```
I think this method is a bit tricky and sometimes hard to follow. Maybe we can start 2 go routines: one writing to a queue and another one checking the results? I think it will simplify the code a lot.

2opremio replied:
I will give it a try. I suspect using two goroutines is going to make synchronization harder.

2opremio replied:
Actually, you were right. I think it's a bit cleaner with 2 subroutines.

2opremio commented on services/horizon/internal/expingest/main.go:
```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+		case <-ps.shutdown:
+			return
+		case reingestRange := <-ps.reingestJobQueue:
+			err := s.ReingestRange(reingestRange.from, reingestRange.to, false)
```
What's the alternative?

2opremio commented on services/horizon/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ file. This project adheres to [Semantic Versioning](http://semver.org/).x
 
 ## Unreleased
 
+* Add `--parallel-workers` to `horizon db reingest range`. `--parallel-workers`  will parallelize reingestion using the supplied number of workers cores to 
```
Huh, odd. Will fix

2opremio commented on services/horizon/internal/expingest/main.go:
```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+	requestedRange ledgerRange
+}
+
+type ParallelSystems struct {
```
I did it, but it's odd that a separate file imports symbols from `main.go`

2opremio commented:
@bartekn PTAL. I addressed your comments, reincorporated `--parallel-job-size` (which will be needed for performance reasons when reingesting the full history) and added an extra test (`TestParallelReingestRangeError()`)

tamirms commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,156 @@
+	wait.Add(1)
+	defer func() {
+		close(stop)
+		wait.Wait()
```
why is it necessary to wait after we have closed `stop`?

2opremio replied:
Because you don't want to return while the goroutine is still executing (it could still be in the middle of a job when stop is closed)

tamirms commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,157 @@
+
+func (ps *ParallelSystems) work(s System) {
+	defer func() {
+		s.Shutdown()
```
is it necessary to call `s.Shutdown()` it seems like the only way `work()` can exit is if the shutdown channel is closed

2opremio replied:
@tamirms That is a call to shut down the underlying dedicated system of the worker (it's just a cleanup call). Are we referring to the same thing?

tamirms replied:
oh sorry, I was confusing the `ParallelSystems` instance with the `System` instance

2opremio replied:
No worries!

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,157 @@
+				// TODO: give account of what ledgers were correctly reingested?
+				errMsg := fmt.Sprintf("in subrange %d to %d",
+					subRangeResult.requestedRange.from, subRangeResult.requestedRange.to)
+				return errors.Wrap(subRangeResult.err, errMsg)
```
This is not correct, it doesn't wait for pending jobs to be finished. @bartekn I will see if I can fix it elegantly but I think it will complicate things enough to restore my original solution.

2opremio replied:
I really don't think it's simpler anymore

bartekn commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	return firstErr
+}
```
I think we can simplify it more:
* A single shutdown signal should be enough to handle everything.
* We don't really need `reingestJobResult` this can be handled inside a worker function.
* We don't need _global_ wait group. `Shutdown` should just sent a shutdown signal and make `ReingestRange` method to wrap up.
* We don't need to start workers when creating a method. Maybe we'll need workers in other methods than `ReingestRange` but now we don't.

After implementing the changes I have the following code:
```go
func (ps *ParallelSystems) ReingestRange(fromLedger, toLedger uint32, batchSizeSuggestion uint32) error {
    var reingestJobQueue = make(chan ledgerRange)
    var erroredMutex sync.Mutex
    var errored bool

    func markError() {
        erroredMutex.Lock()
        errored = true
        erroredMutex.Unlock()
    }

    wg.Add(1)
    // can be moved to a method
    go func() {
        defer wg.Done()
        for subRangeFrom := fromLedger; subRangeFrom < toLedger; {
            // job queuing
            subRangeTo := subRangeFrom + (batchSize - 1) // we subtract one because both from and to are part of the batch
            if subRangeTo > toLedger {
                subRangeTo = toLedger
            }
            select {
            case <-ps.shutdown:
                return
            case reingestJobQueue <- ledgerRange{subRangeFrom, subRangeTo}:
            }
            subRangeFrom = subRangeTo + 1
        }
    }()

    for i := 0; i < workers; i++ {
        wg.Add(1)
        // can be moved to a method
        go func() {
            defer wg.Done()
            s, err := systemFactory(config)
            if err != nil {
                log.Error("...")
                s.Shutdown()
                markError()
                return
            }
            for {
                select {
                case <-ps.shutdown:
                    return
                case reingestRange := <-reingestJobQueue:
                    err := s.ReingestRange(reingestRange.from, reingestRange.to, false)
                    if err != nil {
                        log.Error("...")
                        s.Shutdown()
                        markError()
                        return
                    }
                }
            }
        }
    }

    wg.Wait()
    close(reingestJobQueue)

    if errored {
        return errors.New("one or more jobs failed")
    }
    return nil
}

func (ps *ParallelSystems) shutdown() {
    close(ps.shutdown)

}

// Close should be called when reading is finished.
func (ps *ParallelSystems) Shutdown() error {
	ps.shutdownOnce.Do(msr.shutdown)
	return nil
}

```

2opremio replied:
> * A single shutdown signal should be enough to handle everything.
Not if you have persistent workers

> * We don't really need `reingestJobResult` this can be handled inside a worker function.

Your example has oversimplified error management. If you want to give account of up until what point things where ingested properly (which I will implement shortly after this PR is merged), you need to know what ranges where successful.

> * We don't need _global_ wait group. `Shutdown` should just sent a shutdown signal and make `ReingestRange` method to wrap up.

Not if you have persistent workers

> * We don't need to start workers when creating a method. Maybe we'll need workers in other methods than `ReingestRange` but now we don't.

I did it this way to that we can parallelize stress testing and normal history reingestion later on. 

> After implementing the changes I have the following code

I am not sure how `msr` or `shutdownOnce` are handled. Also, error management is oversimplified. As I mentioned above, you do need to consume the results to know what ranges where processed.

2opremio replied:
@bartekn and I agreed that he will commit his suggestion and we will take it from there.

bartekn commented:
OK, I pushed my code as discussed with @fons. I haven't updated tests, will do it after 👍 from you. I also added a code that return a suggested range to restart a job in case of failure. There's a comment above `lowestRangeErr` explaining why (I think) it works without having to keep track of results from all the jobs.

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,162 @@
+
+}
+
+func (ps *ParallelSystems) Shutdown() error {
```
I don't think it makes sense to have this anymore

2opremio replied:
In fact, having the ability to shutdown an in-progress reingestion will break error reporting (there won't be any guarantee of the lowest failing ledger range being reported)

bartekn replied:
Closing shutdown channel won't abort running jobs. If there's a running `ReingestRange` method in any worker it will complete (succeed or fail but will complete). Also, because the queue channel is unbuffered you won't have any remaining jobs in a buffer (the go routine adding new jobs will return due to `<- ps.shutdown`). Even if this was true, all jobs will complete because there's a wait group that blocks until all go routines return.

Can you elaborate?

2opremio replied:
I think you are right, but I believe it's a bit brittle.

Leaving that aside, I think we agree that `Shutdown()` is not needed anymore since there is no need to cancel running operations or cleanup anything.

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,162 @@
+					}
+					lowestRangeErrMutex.Unlock()
+				}
+				ps.Shutdown()
```
It's awkward that an operation shutdowns the whole `ParallelSystems`. I would remove the global shutdown channel, create one here and pass it to the workers.

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,162 @@
+	}, nil
+}
+
+func (ps *ParallelSystems) reingestWorker(reingestJobQueue <-chan ledgerRange) error {
```
I would use a verb `reingestWorker` (e.g. `runReingestWorker()` `doReingestWork()`)

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,162 @@
+	)
+
+	wg.Add(1)
+	go func() {
```
I don't think there is a reason to do this in a separate goroutine. You could run the inner for loop after spawning the workers.

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,162 @@
+		// Because of this when we reach `wg.Wait()` all jobs previously sent to a channel are processed (either success
+		// or failure). In case of a failure we save the range with the smallest sequence number because this is where
+		// the user needs to start again to prevent the gaps.
+		lowestRangeErr *rangeError
```
this doesn't need to be a pointer (no strong opinions)

bartekn replied:
If it wasn't a pointer we'd need another variable to determine if there was an error or not.

2opremio replied:
you can use the error inside the `rangeError` (if it's nil then no error happened). But, as I said, no strong opinions here.

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,162 @@
+			err := ps.reingestWorker(reingestJobQueue)
+			if err != nil {
+				log.WithError(err).Error("error in reingest worker")
+				if rangeErr, ok := err.(rangeError); ok {
```
This casting isn't necessary if the worker always returns a range error. If you want to distinguish system creation errors, just create the system here.

2opremio replied:
It will also allow to always report the range in which an error happens

bartekn commented:
@fons thanks for a quick review! I added the requested changes and fixed tests. It's great that we're at the same page when it comes to aborting jobs.

I removed `Shutdown` as requested: I agree it doesn't make sense to keep it now because we're not handling system shutdown signals at all in `db reingest` commands. However, I think it's worth doing this in a future so that users get a status update when they terminate the process manually.

I won't be available in the morning tomorrow so please update the code if there's something that's blocking you from merging it.

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,159 @@
+			return nil
+		case reingestRange, more := <-reingestJobQueue:
+			if !more {
+				// Channel closed - no more jobs
```
There is no need to do this, we have the stop channel for that

bartekn replied:
Good point!

2opremio commented on services/horizon/internal/expingest/parallel.go:
```diff
@@ -0,0 +1,158 @@
+				log.WithError(err).Error("error in reingest worker")
+				if rangeErr, ok := err.(rangeError); ok {
+					lowestRangeErrMutex.Lock()
+					if lowestRangeErr == nil || (lowestRangeErr != nil && lowestRangeErr.ledgerRange.from > rangeErr.ledgerRange.from) {
```
There is no need to do `lowestRangeErr != nil ` in the second branch of `||`

2opremio commented:
I think we are done. Thie PR targets the 1.5.0 branch, but my plan is to hold off from merging to avoid disrupting the 1.5.0 release, and just wait until 1.5.0 is merged into master.

But, if @ire-and-curses / @abuiles are OK with it, I will merge now.

ire-and-curses commented:
Let's target this for 1.6 release.

bartekn commented:
Added successful range logging in https://github.com/stellar/go/pull/2724/commits/b0205c909ed35d937024938a75383fd635585474.
